03-rnaseq/pathway-analysis_rnaseq_03_gsva.Rmd

```diff
@@ -435,12 +435,213 @@ sum(duplicated(mapped_df$entrez_id))
+  # Reorder the columns so `gene_means` column is upfront
+  dplyr::select(Ensembl, entrez_id, gene_means, dplyr::everything())
+```
+
+Now we can filter out the duplicate gene identifiers using the gene mean values.
+If we do an `dplyr::arrange()` by `gene_means` first, then this means the first row encountered by `dplyr::distinct()` will be the one with the higher gene mean and by default the first row of any duplicates is kept. 
+
+```{r}
+filtered_mapped_df <- mapped_df %>%
+  # Sort so that the highest mean expression values are at the top
+  dplyr::arrange(dplyr::desc(gene_means)) %>%
+  # Filter out the duplicated rows using `dplyr::distinct()`
+  dplyr::distinct(entrez_id, .keep_all = TRUE)
+```
+
+Let's do our check again to see if we still have duplicates
+
+```{r}
+sum(duplicated(mapped_df$entrez_id))
+```
+
+We now have `0` duplicates which is what we want. All set!
+
+Now we should prep this data so GSVA can use it. 
+
+```{r}
+filtered_mapped_matrix <- filtered_mapped_df %>%
+  # GSVA can't the Ensembl IDs so we should drop this column as well as the means
+  dplyr::select(-Ensembl, -gene_means) %>%
+  # We need to store our gene identifiers as row names
+  tibble::column_to_rownames("entrez_id") %>%
+  # Now we can convert our object into a matrix
+  as.matrix()
+```
+
+Note that if we had duplicate gene identifiers here, we would not be able to set them as row names.
+
+## GSVA - Microarray
+
+Rather than ranking genes based on some gene-level statistic selected by the user, GSVA fits a model and ranks genes based on their expression level relative to the sample distribution.
+The pathway-level score calculated is a way of asking how genes _within_ a gene set vary as compared to genes that are _outside_ of that gene set [@Malhotra2018].
+
+The idea here is that we will get pathway-level scores for each sample that indicate if genes in a pathway vary concordantly in one direction (over-expressed or under-expressed relative to the overall population) [@Hanzelmann2013].
+
+The output is a gene set by sample matrix of GSVA scores.
+
+### Perform GSVA
+
+Let's perform GSVA using the `gsva()` function.
+See `?gsva` for more options.
+
+```{r}
+gsva_results <- gsva(
+  filtered_mapped_matrix,
+  hallmarks_list,
+  method = "gsva",
+  # Appropriate for our normalized counts data
+  kcdf = "Gaussian",
+  # Minimum gene set size
+  min.sz = 15,
+  # Maximum gene set size
+  max.sz = 500,
+  # Compute Gaussian-distributed scores
+  mx.diff = TRUE,
+  # Don't print out the progress bar
+  verbose = FALSE
+)
+```
+
+Note that the `gsva()` function documentation says we can use `kcdf = "Gaussian"` if we have expression values that are continuous such as log-CPMs, log-RPKMs or log-TPMs, but we would use `kcdf = "Poisson"` on integer counts.
+Our `vst()` transformed data is on a log2-like scale, so `Gaussian` works for us. 
+
+Let's explore what the output of `gsva()` looks like.
+
+```{r}
+head(gsva_results[, 1:10])
+```
+
```

Comment:
This section is almost identical to the microarray example. Just the one sentence about vst is added.

Reply:
I assume we want this note in the microarray example?

Reply:
No, I asked for it to come out because it's less relevant.

Reply:
AH. Okay. Right. So it's NOT there and we don't want it to be. :+1: